Implement --target_pattern_file flag for reading patterns from file

robbertvanginkel opened the pull request:
An alternative to https://github.com/bazelbuild/bazel/pull/10796 to fix https://github.com/bazelbuild/bazel/issues/8609.

As https://github.com/bazelbuild/bazel/issues/8609#issuecomment-590522448 points out there's currently a `--query_file` to read a query pattern from file, but there's no analogous feature available for build/test. Generic parameter file support attempted in https://github.com/bazelbuild/bazel/pull/10796 turns out to be a little more complex than expected due to bazel's client/server architecture and likely requires some design document. 

This PR would fix help https://github.com/bazelbuild/bazel/issues/8609 without rushing into any designs, so we can spend time properly designing generic parameter files while also alleviating the immediate problem with a  known/simpler pattern.

robbertvanginkel commented:
@michajlo thanks for your comments in https://github.com/bazelbuild/bazel/pull/10796#pullrequestreview-363636974 and your suggestion in https://github.com/bazelbuild/bazel/issues/8609#issuecomment-590522448. While would like generic parameter file support in Bazel, I would also appreciate a more immediate solution to #8609. Suggesting we go with this much simpler approach for now till I can spend the time on a generic solution to the open questions in https://github.com/bazelbuild/bazel/pull/10796.

robbertvanginkel commented on src/main/java/com/google/devtools/build/lib/runtime/commands/BuildCommand.java:
```diff
@@ -94,4 +100,31 @@ public BlazeCommandResult exec(CommandEnvironment env, OptionsParsingResult opti
     return BlazeCommandResult.exitCode(exitCode);
   }
+
+  public static Pair<ExitCode, List<String>> getTargets(CommandEnvironment env, OptionsParsingResult options) {
```
Pairs with `null` are probably not the most elegant way of reusing this code between `BuildCommand`/`TestCommand`, suggestions to improve welcome :)

michajlo replied:
Might be cleaner to move to a new class and use a custom exception to signal that whatever was invalid?

michajlo commented on src/main/java/com/google/devtools/build/lib/buildtool/BuildRequestOptions.java:
```diff
@@ -455,6 +455,17 @@ public boolean useTopLevelTargetsForSymlinks() {
   public boolean replayActionOutErr;
 
+  @Option(
+      name = "build_file",
```
let's call this different - "build file" is already claimed... maybe "target_file" or "targets_file" or "target_pattern_file"?

robbertvanginkel replied:
Went with `target_pattern_file`.

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/commands/BuildCommand.java:
```diff
@@ -64,12 +69,29 @@
+      // Works for absolute or relative query file.
+      Path residuePath = env.getWorkingDirectory().getRelative(buildRequestOptions.buildFile);
+      try {
+        targets = Lists.newArrayList(FileSystemUtils.readLines(residuePath, StandardCharsets.UTF_8));
```
We should probably check if not targets were specified in the files. If we're not doing this in --query_file it was probably an oversight...

robbertvanginkel replied:
The build command explicitly allows calling build without a resisdue/target pattern.

michajlo replied:
I think this was mostly resolved by later commits - IIRC my concern was the target file could be empty and we'd be skip the "Nothing will be built" warning.

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/commands/BuildCommand.java:
```diff
@@ -69,29 +70,12 @@
-    } else {
+    Pair<ExitCode, List<String>> maybeTargets = getTargets(env, options);
+    if (maybeTargets.getFirst() != null) {
+      BlazeCommandResult.exitCode(maybeTargets.getFirst());
```
did you mean to return this?

michajlo commented on src/test/shell/integration/build_file_test.sh:
```diff
@@ -0,0 +1,102 @@
+
+function test_build_file_no_targetpattern() {
+  setup
+  bazel build --build_file=build.params -- //:x >& $TEST_log && fail "expected failure" || true
```
... && fail "Expected failure"

We want to make sure this command didn't succeed.

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupport.java:
```diff
@@ -0,0 +1,50 @@
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+public final class TargetPatternFileSupport {
```
make package private (remove "public")? Also add a quick javadoc to this and the visible methods and exception.

jin commented:
Reviewers: once approved, please read the internal go/bazel-import-pr link for PR import instructions.

robbertvanginkel commented:
Processed the comments but did not get to the unit test yet. I looked at making one, but standing up a `CommandEnvironment` to pass in is a little tricky (its constructor is not public and in a different package), and when I attempted to make a mockito mock I got an error like 
```
Cannot mock/spy class com.google.devtools.build.lib.runtime.CommandEnvironment
Mockito cannot mock/spy because :
 - final class
```
I'm not sure how best to approach that, if you have any recommendation please let me know and I'll try to finish the unit test.

michajlo commented:
CommandEnvironment should be mockable as of an hour or so ago.

michajlo commented:
Let me know when that last test is ready, otherwise everything else lgtm.

robbertvanginkel commented:
@michajlo the unit test should be there: https://github.com/bazelbuild/bazel/blob/20964788553921e0335d5c34cd4bf70d18c4b7dd/src/test/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupportTest.java

robbertvanginkel commented:
@michajlo I squashed the changes and rebased them on master to resolve a merge conflict github was showing. It seems like there was a few changes to use `DetailedExitCode` instead of `ExitCode`, I attempted to update the references.

I think all the changes, including unit and integration tests should be there. Please let me know if theres anything else!

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupport.java:
```diff
@@ -0,0 +1,63 @@
+   * list of targets based on the supplied command line options.
+   */
+  public static class TargetPatternFileSupportException extends Exception {
+    public TargetPatternFileSupportException(String message) { super(message); }
```
style thing, mv the super(...) to a new line:

    ... {
      super(message);
    }

robbertvanginkel replied:
Fixed. Does Bazel have something like `google-java-format` setup? I couldn't find it myself in the contribution guidelines.

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupport.java:
```diff
@@ -0,0 +1,63 @@
+   * delimited target patterns from the --target_pattern_file flag. If --target_pattern_file is
+   * specified and options contain a residue, or file cannot be read it throws an exception instead.
+   */
+  public static List<String> handleTargetPatternFile(
```
nit: this isn't always getting the target patterns from a file, in fact it's almost always going to come from the commandline. Can we rename this to better reflect that? I think renaming the class/method to something like "TargetPatternsHelper#readFrom" avoids committing to too much.

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupport.java:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  /**
+   * TargetPatternFileSupportException gets thrown when TargetPatternFileSupport cannot return a
```
I think you can just simplify this to "Thrown when target patterns were incorrectly specified" since it can happen independent of any file being read.

michajlo commented on src/test/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupportTest.java:
```diff
@@ -0,0 +1,103 @@
+        TargetPatternFileSupport.TargetPatternFileSupportException.class,
+        () -> TargetPatternFileSupport.handleTargetPatternFile(env, options));
+  }
+}
```
add a test that specifying neither is ok?

robbertvanginkel commented:
All comments should be taken care of.

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupport.java:
```diff
@@ -0,0 +1,63 @@
+package com.google.devtools.build.lib.runtime.commands;
```
this file needs a copyright header

robbertvanginkel commented:
Thanks @michajlo!

isaprykin commented:
--query_file accepts space separated list of targets.
It needs to be clarified that this only accepts newline separated list of targets.
I costed me a few hours.
